src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -1156,6 +1175,106 @@ private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, Ty
+                                builder.Add(_factory.Call(dictionaryTemp, addMethod, rewrittenKey, rewrittenValue));
+                                var statements = builder.SelectAsArray(expr => (BoundStatement)new BoundExpressionStatement(expr.Syntax, expr));
+                                builder.Free();
+                                Debug.Assert(statements.Length > 0);
+                                return statements.Length == 1 ?
+                                    statements[0] :
+                                    new BoundBlock(iteratorBody.Syntax, locals: ImmutableArray<LocalSymbol>.Empty, statements);
+                            });
+                        sideEffects.Add(rewrittenElement);
+                        break;
+                    default:
+                        {
+                            var (rewrittenKey, rewrittenValue) = RewriteKeyValuePair((BoundExpression)element, addMethod, sideEffects, localsBuilder);
+                            sideEffects.Add(_factory.Call(dictionaryTemp, addMethod, rewrittenKey, rewrittenValue));
+                        }
+                        break;
+                }
+            }
+
+            var locals = localsBuilder.SelectAsArray(l => l.LocalSymbol);
+            localsBuilder.Free();
+
+            return new BoundSequence(
+                node.Syntax,
+                locals,
+                sideEffects.ToImmutableAndFree(),
+                dictionaryTemp,
+                collectionType);
+        }
+
+        private (BoundExpression, BoundExpression) RewriteKeyValuePair(
+            BoundExpression expression,
+            MethodSymbol addMethod,
+            ArrayBuilder<BoundExpression> sideEffects,
+            ArrayBuilder<BoundLocal> localsBuilder)
+        {
+            Debug.Assert(expression.Type is { });
+
+            var sourceType = (NamedTypeSymbol)expression.Type;
+            if (!ConversionsBase.IsKeyValuePairType(_compilation, sourceType, WellKnownType.System_Collections_Generic_KeyValuePair_KV, out var sourceKeyType, out var sourceValueType))
+            {
+                throw ExceptionUtilities.UnexpectedValue(expression.Type);
+            }
+
+            BoundAssignmentOperator assignmentToTemp;
+            BoundLocal expressionTemp = _factory.StoreToTemp(VisitExpression(expression), out assignmentToTemp);
+            localsBuilder.Add(expressionTemp);
+            sideEffects.Add(assignmentToTemp);
+            var getKeyMethod = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Collections_Generic_KeyValuePair_KV__get_Key)).AsMember(sourceType);
+            var getValueMethod = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Collections_Generic_KeyValuePair_KV__get_Value)).AsMember(sourceType);
+            return (_factory.Convert(addMethod.Parameters[0].Type, _factory.Call(expressionTemp, getKeyMethod)),
```

Comment:
This seems wrong. We're making brand new conversions here and they are not the same that we checked and calculated during initial binding (those were attempts to convert the whole element, but here we're converting the key and the value separately)